[RFC] vim-patch:8.1.{0309,0362,0365,0515,1946}

erw7 opened the pull request:
**vim-patch:8.1.0362: cannot get the script line number when executing a function**
Problem:    Cannot get the script line number when executing a function.
Solution:   Store the line number besides the script ID. (Ozaki Kiichi, closes vim/vim#3362)  Also display the line number with ":verbose set".
https://github.com/vim/vim/commit/f29c1c6aa3f365c025890fab5fb9efbe88eb1761

**vim-patch:8.1.0309: profiling does not show a count for condition lines**
Problem:    Profiling does not show a count for condition lines.  (Daniel Hahler)
Solution:   Count lines when not skipping. (Ozaki Kiichi, closes #2499)
vim/vim@7feb35e

**vim-patch:8.1.0365: function profile doesn't specify where it was defined**
Problem:    Function profile doesn't specify where it was defined.
Solution:   Show the script name and line number.
vim/vim@4c7b08f

**vim-patch:8.1.0515: reloading a script gives errors for existing functions**
Problem:    Reloading a script gives errors for existing functions.
Solution:   Allow redefining a function once when reloading a script.
vim/vim@ded5f1b

**vim-patch:8.1.1946: memory error when profiling a function without a script ID**
Problem:    Memory error when profiling a function without a script ID.
Solution:   Check for missing script ID.  (closes vim/vim#4877)
vim/vim@1635880

erw7 committed:
vim-patch:8.1.0362: cannot get the script line number when executing a function

Problem:    Cannot get the script line number when executing a function.
Solution:   Store the line number besides the script ID. (Ozaki Kiichi,
            closes vim/vim#3362)  Also display the line number with ":verbose set".
https://github.com/vim/vim/commit/f29c1c6aa3f365c025890fab5fb9efbe88eb1761

erw7 committed:
Change test because maparg was changed to also return lnum

erw7 committed:
[squash] Fix errors when porting

justinmk commented:
@erw7 thanks, this is appreciated! @abdelhakeem will this conflict with your PR?

abdelhakeem commented:
> will this conflict with your PR?

Yeah, no problem, I'll rebase and resolve it before pushing some other local changes

erw7 committed:
vim-patch:8.1.0365: function profile doesn't specify where it was defined

Problem:    Function profile doesn't specify where it was defined.
Solution:   Show the script name and line number.
https://github.com/vim/vim/commit/4c7b08f6409349cf5324a26f7523c438590550ef

erw7 committed:
vim-patch:8.1.0515: reloading a script gives errors for existing functions

Problem:    Reloading a script gives errors for existing functions.
Solution:   Allow redefining a function once when reloading a script.
https://github.com/vim/vim/commit/ded5f1bed7ff2d138b3ee0f9610d17290b62692d

erw7 committed:
vim-patch:8.1.0309: profiling does not show a count for condition lines

Problem:    Profiling does not show a count for condition lines.  (Daniel
            Hahler)
Solution:   Count lines when not skipping. (Ozaki Kiichi, closes #2499)
https://github.com/vim/vim/commit/7feb35e7782907b44659a2748ff5d7489deeed74

erw7 committed:
vim-patch:8.1.1946: memory error when profiling a function without a script ID

Problem:    Memory error when profiling a function without a script ID.
Solution:   Check for missing script ID.  (closes vim/vim#4877)
https://github.com/vim/vim/commit/163588005da3a240e49416093d0d0251951d60a1

erw7 commented:
Since https://github.com/neovim/neovim/pull/10869/commits/c005c4b727bc995db95edb030b0cf3cb1a9a0caf is required by vim, vim/vim#4877 was created.

janlazo commented on src/nvim/eval.c:
```diff
@@ -22041,16 +22041,18 @@ void func_dump_profile(FILE *fd)
-                p, fp->uf_script_ctx.sc_lnum);
-        if (should_free) {
-          xfree(p);
+        if (fp->uf_script_ctx.sc_sid != 0) {
```
[vim-patch:8.1.1946](https://github.com/vim/vim/commit/163588005da3a240e49416093d0d0251951d60a1)

erw7 replied:
Changes to `test_profile.vim` in [vim-patch:8.1.1946](https://github.com/vim/vim/commit/163588005da3a240e49416093d0d0251951d60a1) cannot be imported due to lack of patches. Isn't it necessary to reflect changes to `test_profile.vim` because it is a test that is not run?

janlazo replied:
That test is N/A so you can omit the test. Comment out the line, `source screendump.vim`.

justinmk replied:
We could have an empty `screendump.vim` so that such lines don't need to be commented-out.

justinmk commented on src/nvim/globals.h:
```diff
@@ -332,8 +332,8 @@ EXTERN int garbage_collect_at_exit INIT(= false);
-// ID of script being sourced or was sourced to define the current function.
-EXTERN scid_T current_SID INIT(= 0);
+// Script CTX being sourced or was sourced to define the current function.
+EXTERN sctx_T current_sctx INIT(= { 0 COMMA 0 });
```
I wonder if we actually need `COMMA`. It's from https://github.com/vim/vim/commit/ac77aec4daea8d73468fcf4690cb4ccab1d807ed . 

We use `,` (literal comma) when defining `g_stats`, and we don't get a clang warning.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -1498,22 +1482,61 @@ static char_u * do_one_cmd(char_u **cmdlinep,
+  //
+  // The ea.cmd pointer is updated to point to the first character following the
+  // range spec. If an initial address is found, but no second, the upper bound
+  // is equal to the lower.
```
lost some formatting here. I'll fix it in the merge.

justinmk commented:
@erw7  thank you!
